File: ipc_reader.cpp
#include <windows.h>
#include <iostream>

int main() {
    HANDLE hFile = CreateFileA("ipc_queue.bin", GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
    if (hFile == INVALID_HANDLE_VALUE) {
        std::cout << "File not found. Run Java producer first.\n";
        return 1;
    }

    HANDLE hMap = CreateFileMappingA(hFile, NULL, PAGE_READONLY, 0, 0, NULL);
    void* pBuf = MapViewOfFile(hMap, FILE_MAP_READ, 0, 0, 0);

    long long* dataPtr = static_cast<long long*>(pBuf);
    long long currentIndex = dataPtr[0];

    std::cout << "C++ HFT Consumer Start\n";
    std::cout << "Java Queue Current Index: " << currentIndex << "\n";

    if (currentIndex > 0) {
        long long targetIndex = (currentIndex - 1) % 10000;
        long long lastData = dataPtr[1 + targetIndex];
        std::cout << "Intercepted Data from Java: " << lastData << "\n";
    }

    UnmapViewOfFile(pBuf);
    CloseHandle(hMap);
    CloseHandle(hFile);

    return 0;
}

File: ipc_reader.py
import mmap
import struct
import os

def main():
    file_path = "ipc_queue.bin"

    if not os.path.exists(file_path):
        print("Error: ipc_queue.bin not found. Run Java producer first.")
        return

    with open(file_path, "r+b") as f:
        mm = mmap.mmap(f.fileno(), 0)

        header_bytes = mm[0:8]
        current_index = struct.unpack("<q", header_bytes)[0]

        print("Polyglot IPC Consumer Start (Python)")
        print("Java Queue Current Index:", current_index)

        if current_index > 0:
            target_index = (current_index - 1) % 10000
            offset = 8 + (target_index * 8)

            data_bytes = mm[offset:offset+8]
            last_data = struct.unpack("<q", data_bytes)[0]

            print("Intercepted Data from Java:", last_data)

        mm.close()

if __name__ == "__main__":
    main()